### PR TITLE
SpeziQuestionnaire

### PR DESCRIPTION
**Pull request by pauljohanneskraft**
# SpeziQuestionnaire

## :recycle: Current situation & Problem
*Link any open issues or pull requests (PRs) related to this PR. Please ensure that all non-trivial PRs are first tracked and discussed in an existing GitHub issue or discussion.*

## :gear: Release Notes 
*Add a bullet point list summary of the feature and possible migration guides if this is a breaking change so this section can be added to the release notes.*
*Include code snippets that provide examples of the feature implemented or links to the documentation if it appends or changes the public interface.*

## :books: Documentation
*Please ensure that you properly document any additions in conformance to [Spezi Documentation Guide](https://github.com/StanfordSpezi/.github/blob/main/DOCUMENTATIONGUIDE.md).*
*You can use this section to describe your solution, but we encourage contributors to document your reasoning and changes using in-line documentation.* 

## :white_check_mark: Testing
*Please ensure that the PR meets the testing requirements set by CodeCov and that new functionality is appropriately tested.*
*This section describes important information about the tests and why some elements might not be testable.*

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pauljohanneskraft** [comment]
I unfortunately didn't manage to get the previews and tests to work with this and didn't have the time yet to further investigate. Since the previous implementation also didn't include tests for this part, I assume that it isn't easily possible...

**Comment by pauljohanneskraft** [comment]
The failing tests seem to be related to the new dependencies, since I didn't want to put my files in the app-module, so ready for review anyways and either please let me know how to fix this or I will just put them somewhere else.

After merging all the open PRs, we should think about the reorganization anyways probably, but I'm not sure, whether I want to do that before submission

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/spezi/QuestionnaireComposable.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.Questionnaire
+
+@Composable
+fun QuestionnaireComposable(
+    questionnaire: Questionnaire,
+    modifier: Modifier = Modifier,
+    completionStepMessage: String? = null,
+    cancelBehavior: CancelBehavior = CancelBehavior.ShouldConfirmCancel,
+    onResult: suspend (QuestionnaireResult) -> Unit,
```
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/spezi/QuestionnaireComposable.kt:29]: There should be no need to have this suspendable. We just notify the callback, and clients can use their own scope (not our composable `rememberCoroutineScope()`) to process the result. See also usage in Engage that there is indeed no need to have this suspend, and in general no composable callbacks

File: app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/spezi/QuestionnaireComposable.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            }
+        }
+    }
+    val fullQuestionnaire = remember(questionnaire, completionItem) {
+        completionItem?.let {
+            questionnaire.copy().apply { addItem(it) }
+        } ?: questionnaire
+    }
+    val questionnaireJson = remember(fullQuestionnaire) {
+        FhirContext.forR4()
```
Comment (eldcn) [inline app/src/main/kotlin/edu/stanford/bdh/engagehf/questionnaire/spezi/QuestionnaireComposable.kt:46]: encoding can fail, I would avoid crashing at this level, and use:

```kotlin
val questionnaireJson = remember(fullQuestionnaire) {
    runCatching {
        FhirContext.forR4()
            .newJsonParser()
            .encodeResourceToString(fullQuestionnaire)
    }.getOrDefault("")
}
```

We can delegate the invalid json error screen to the library

**Review by eldcn — APPROVED** [review]
Looks good for now and from my side we can merge :) As discussed, we should consider in a later step restructuring into separate modules

**Comment by pauljohanneskraft** [comment]
Due to the missing code coverage reporting in that package, you will probably need to merge without that check - would that work and is my assumption right? Since the test should actually cover that code.... @eldcn

**Pull request merged**